### ISSUE TITLE
Nvidia distrobox integration not working

### ISSUE THREAD

**Issue by KyleGospo** [ublue-os/bazzite#61]
Distrobox with the `--nvidia` flag is supposed to copy the host's Nvidia drivers into the container. It seems this may be incomplete, as currently this results in a broken installation.

This is a blocker for desktop 1.0

**Comment by KyleGospo** [comment]
https://github.com/89luca89/distrobox/issues/848

This may be a hint as to what to look for.

**Comment by bsherman** [comment]
> this results in a broken installation.

what does "broken installation" mean specifically?

**Comment by KyleGospo** [comment]
In this case non-functioning Vulkan within the container, possibly OpenGL as well.

We tested both --nvidia and the launch options to use the Nvidia Container Toolkit documented in distrobox's readme

**Comment by bsherman** [comment]
Here's some basic feedback.

Host is an Intel i5-8400 desktop with iGPU disabled and NVIDIA 1660 Ti as primary/only GPU:

```
[bsherman@i5nvidia ~]$ rpm-ostree status -b
State: idle
AutomaticUpdates: stage; rpm-ostreed-automatic.service: last run failed
BootedDeployment:
● ostree-unverified-image:docker://ghcr.io/ublue-os/silverblue-nvidia:38
                   Digest: sha256:72fa0a0ac75563ed0161669e48e16289c898b90861d91c2a43c3bc5bc61bd6e0
                  Version: 38.20230724.0 (2023-07-24T21:04:30Z)
```

This is a pretty clean install, I did it just to test the installer once and then did an update yesterday. So distrobox is from Fedora repos.

Created 2 distroboxes like so:
```
distrobox create --nvidia --name fedora-nvidia 
distrobox create --nvidia --name ubuntu-nvidia --image ubuntu:latest
```

and this much works:
```
[bsherman@i5nvidia ~]$ distrobox enter fedora-nvidia
[bsherman@fedora-nvidia ~]$ nvidia-smi 
Tue Jul 25 11:17:31 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1660 Ti     Off | 00000000:01:00.0 Off |                  N/A |
|  0%   35C    P8              10W / 130W |    162MiB /  6144MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1899      G   /usr/bin/gnome-shell                        112MiB |
+---------------------------------------------------------------------------------------+
```

```
[bsherman@i5nvidia ~]$ distrobox enter ubuntu-nvidia
bsherman@ubuntu-nvidia:~$ nvidia-smi 
Tue Jul 25 11:17:57 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1660 Ti     Off | 00000000:01:00.0 Off |                  N/A |
|  0%   35C    P8              10W / 130W |    162MiB /  6144MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1899      G   /usr/bin/gnome-shell                        112MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by KyleGospo** [comment]
Looks like this may have been fixed recently then, that's great!

As a final test, could you try bazzite-arch in distrobox?
`distrobox create --name bazzite-arch --image ghcr.io/ublue-os/bazzite-arch:latest --nvidia`

**Comment by bsherman** [comment]
I can confirm that the vulkan bug linked still seems to exist on my current test intallation

> 89luca89/distrobox#848

**Comment by bsherman** [comment]
I tested with this 
> distrobox create --name bazzite-arch --image ghcr.io/ublue-os/bazzite-arch:latest --nvidia
and with
> distrobox create --name bazzite-arch --image ghcr.io/ublue-os/bazzite-arch:latest --nvidia --volume /etc/vulkan/:/etc/vulkan/ --volume /usr/lib64/libnvidia-vulkan-producer.so:/usr/lib64/libnvidia-vulkan-producer.so

in an attempt to get vulkan working like in the linked ticket.

In short, OpenGL (glxgears/glxinfo) works fine for me in the bazzite-arch distrobox. I didn't get vulkan working with either of those different distrobox setups.

**Comment by bsherman** [comment]
For good measure I just did the same tests with bazzite-arch on my laptop (hybrid intel + nvidia 3050ti ) and I got the same results as the i5 desktop + 1660ti test.

**Comment by bsherman** [comment]
another test to check that `libnvidia-vulkan-producer.so` on fedora...

```
[bsherman@i5nvidia ~]$ distrobox create --nvidia --name fedora-nvidia  
Creating 'fedora-nvidia' using image registry.fedoraproject.org/fedora-toolbox:38	[ OK ]
Distrobox 'fedora-nvidia' successfully created.
To enter, run:

distrobox enter fedora-nvidia

[bsherman@i5nvidia ~]$ distrobox create --nvidia --name fedora-nvidia2 --volume /usr/lib64/libnvidia-vulkan-producer.so:/usr/lib64/libnvidia-vulkan-producer.so
Creating 'fedora-nvidia2' using image registry.fedoraproject.org/fedora-toolbox:38	[ OK ]
Distrobox 'fedora-nvidia2' successfully created.
To enter, run:

distrobox enter fedora-nvidia2
```
and then check for presence of that shared object

```
[bsherman@i5nvidia ~]$ distrobox enter fedora-nvidia
[bsherman@fedora-nvidia ~]$ ls -l /usr/lib64/libnvidia-vulkan-producer.so
ls: cannot access '/usr/lib64/libnvidia-vulkan-producer.so': No such file or directory
[bsherman@fedora-nvidia ~]$ exit
logout
[bsherman@i5nvidia ~]$ distrobox enter fedora-nvidia2
[bsherman@fedora-nvidia2 ~]$ ls -l /usr/lib64/libnvidia-vulkan-producer.so
-rwxr-xr-x. 3 nobody nobody 22856 Dec 31  1969 /usr/lib64/libnvidia-vulkan-producer.so
[bsherman@fedora-nvidia2 ~]$ exit
logout
```

and... vkcube test from `dnf install vulkan-tools` on both instances

```
[bsherman@fedora-nvidia ~]$ vkcube
Selected GPU 0: NVIDIA GeForce GTX 1660 Ti, type: DiscreteGpu

[bsherman@fedora-nvidia2 ~]$ vkcube
Selected GPU 0: NVIDIA GeForce GTX 1660 Ti, type: DiscreteGpu
```

it works!

**Comment by bsherman** [comment]
Ok, an update... I see successful vulkan  with `vkcube` even in bazzite-arch, but i get a core dump on `vkgears`

```
bazzite-arch in ~ took 2s 
at 12:50:31 ⬢ ❯ vkcube
Selected GPU 0: NVIDIA GeForce RTX 3050 Ti Laptop GPU, type: DiscreteGpu

bazzite-arch in ~ took 3s 
at 12:50:37 ⬢ ❯ vkgears 
vkgears: ../mesa-demos-9.0.0/src/vulkan/vkgears.c:400: create_swapchain: Assertion `surface_caps.supportedCompositeAlpha & VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR' failed.
Aborted (core dumped)
```

**Comment by bsherman** [comment]
Weird though.,.. vulkaninfo NOT working

```
fedora-nvidia in ~ took 2s 
at 14:30:38 ⬢ ❯ vulkaninfo
ERROR at /builddir/build/BUILD/Vulkan-Tools-sdk-1.3.243.0/vulkaninfo/vulkaninfo.h:1475:vkGetPhysicalDeviceSurfaceSupportKHR failed with ERROR_INITIALIZATION_FAILED
```

I'll try it again with "my fix" ( https://github.com/ublue-os/bazzite/commit/91c2b39b5151d3375ccfcef33d22a6f79bfb4c0b )

**Comment by bsherman** [comment]
Nope, i still get that error on this instance
```
distrobox create --nvidia --name fedora-nvidia2 --volume /usr/lib64/libnvidia-vulkan-producer.so:/usr/lib64/libnvidia-vulkan-producer.so

[bsherman@fedora-nvidia2 ~]$ vulkaninfo 
ERROR at /builddir/build/BUILD/Vulkan-Tools-sdk-1.3.243.0/vulkaninfo/vulkaninfo.h:1475:vkGetPhysicalDeviceSurfaceSupportKHR failed with ERROR_INITIALIZATION_FAILED
```

**Comment by KyleGospo** [comment]
Fixed in 91c2b39b5151d3375ccfcef33d22a6f79bfb4c0b
